### PR TITLE
[Core][AnalysisStage] processes add vars => #869

### PR DESCRIPTION
**Pull request by philbucher**
implements #869 

With this PR the processes have the option to specify which historical variables they need 
=> for this they have to implement `GetHistoricalVariables` which is a "static" method to extract the required vars from the parameters

Then the solver has to implement `AddAdditionalVariables` in order to process the additional variables

Also derived classes can use `_GetListOfAdditionalHistoricalVariables` for adding more variables, e.g. for utilities as @maceligueta was mentioning

### CODE REVIEW

**Review by loumalouomega — DISMISSED** [review]
This is duplicated with the solver in StructuralMechanicsApplications. Remove from there allowing a compatible interface

**Comment by philbucher** [comment]
what do you mean is duplicated?

**Comment by loumalouomega** [comment]
> what do you mean is duplicated?

https://github.com/KratosMultiphysics/Kratos/blob/6133c9308f7506bd9604ef9c81698e8cd60fb8d7/applications/StructuralMechanicsApplication/python_scripts/structural_mechanics_solver.py#L89 

here

**Comment by philbucher** [comment]
ah that one
This can be done in a later PR, also we have to deprecate it first

**Comment by loumalouomega** [comment]
> ah that one
> This can be done in a later PR, also we have to deprecate it first

But at least a retrocompatible interface is required, I remove my block

**Comment by RiccardoRossi** [comment]
let s begin with an opinion of the implementation committee 😀

**Comment by loumalouomega** [comment]
I already gave mine

**Comment by RiccardoRossi** [comment]
well, the committee shoukd speak with a single voice...after reaching a
consensus.

the idea is exactly this..when an agreement is reached (strictly by
consensus) you pass your reasoned advice to the technical committee.

File: kratos/python_scripts/analysis_stage.py [hunk 1]
```diff
@@ -274,3 +276,26 @@ def __CreateListOfProcesses(self):
         order_processes_initialization = self._GetOrderOfOutputProcessesInitialization()
         self._list_of_output_processes = self._CreateProcesses("output_processes", order_processes_initialization)
         self._list_of_processes.extend(self._list_of_output_processes) # Adding the output processes to the regular processes
+
+    def _GetListOfAdditionalHistoricalVariables(self):
+        """This function returns the list of (additional) historicsl variables
```
Comment (rubenzorrilla) [inline kratos/python_scripts/analysis_stage.py:281]: Typo

**Comment by philbucher** [comment]
Looking at this again after some time I realized that it is not enough to only pass the list of variables to be added:
Also it is necessary to specify the corresponding ModelPart, e.g. for a coupled solver!
I hadn't thought abt this before.

I will improve this soon, but probably after the release

**Comment by philbucher** [comment]
@maceligueta: @jcotela told me that you would like a method with which you can add variables to the MP from the AnalysisStage

This PR does exactly that (originally intended to be used by the processes, but can also be used for other things I suppose), it is just not ready yet, but if it is urgent for you I can do it soon.

The idea is the following now: the solver gets through`AddAdditionalVariables` a map of MainModelPart-Names each having a list of Variables such that it can add the variables

e.g. 
~~~py
additional_vars = {
    "FluidMP" :    [DUMMY_VELOCITY, AUX_DISTANCE], # to be discussed if passed as strings or directly variables
    "Structure" :  [ROTATION],
    "ParticelMP" : [INTERFACE_FORCE, AUX_ACCELERATION]
}
~~~

is this what you would need?

**Comment by philbucher** [comment]
ping @maceligueta

**Comment by maceligueta** [comment]
I think this PR is totally necessary. However, I think it would be clearer if we dedicate one line for just creating the solver. After that, we could call a function that adds all the variables (through a list returned by the solver and the processes).
> The idea is the following now: the solver gets through `AddAdditionalVariables` a map of MainModelPart-Names each having a list of Variables such that it can add the variables

@philbucher 
The idea of a map is indeed very nice. I understand that each object would return one of those maps, right? The solver, the processes, ... And then the Stage itself (not the Solver) adds all the variables at once? If so, I very much like the idea.

**Comment by philbucher** [comment]
I see where this is going @maceligueta and I like it
I will make a draft and then we can discuss in code, I think it’s easier in this case

**Comment by maceligueta** [comment]
I talked to @RiccardoRossi  today and he is worried that the Solver might need the variables in its constructor. The processes as well, I guess.
If solvers need the variables in the modelpart for their construction but the modelparts don't have those variables yet... we kind of have a deadlock.

**Comment by philbucher** [comment]
Note that the modelpart is read by the time the processes are constructed, that’s why I am introducing a „static“ method for them here
For the solver though this is NOT the case, it is first constructed and then later the modelpart is read.
Also since the solver does know which variables it uses I don’t fully understand what could be the problem.
@RiccardoRossi could you please explain more in detail what is your concern?

**Comment by RiccardoRossi** [comment]
hi philipp,

essentially i am not sure of what is the irder at ehich you would call things.

we have a factory method for the construction of solvers but nothing of the type for the variables, so i am not clear of how this would work. same applies for the processes.

to make it clear, i am not against your oriposal, just i am not ckear of hiw to make it to work

**Comment by ddiezrod** [comment]
@philbucher @RiccardoRossi Is this PR still alive? I find this quite useful

**Comment by philbucher** [comment]
It is still alive, I just did not have time to continue, sorry for the delay, will try to finish it soon

**Comment by philbucher** [comment]
Closing this for the moment as it is heavily outdated and can probably be done better by now

Sorry for the inconveniences, if anyone really needs this we can discuss to reopen

**Comment by ddiezrod** [comment]
Hi @philbucher I definitely need this for one of my developments. I think the main idea of this is correct. I'd finish it myself but I am not sure what limitations you found here to close this. Could you please share your thoughts?

**Comment by philbucher** [comment]
Hey @ddiezrod 
The issue is a bit of a chicken-egg problem. The historical variables need to be added to the ModelParts before reading the mdpas, but the processes can only be constructed after the mdpas are read.
Hence you need some sort of static method for the process, which can tell you which vars to add to which ModelPart, but based ONLY on the settings, which might be limiting a bit in some cases.

Also as you can see my implementation was a bit messy, you can probably do it better if you add such a method in the Cpp Process

Let me know if still sth is unclear or you have more questions

**Comment by loumalouomega** [comment]
FYI: https://github.com/KratosMultiphysics/Kratos/blob/master/kratos/tests/test_specifications_utilities.py

**Comment by loumalouomega** [comment]
This is not yet added to the analysis/solvers

**Comment by ddiezrod** [comment]
> Hence you need some sort of static method for the process, which can tell you which vars to add to which ModelPart, but based ONLY on the settings, which might be limiting a bit in some cases.

Honestly I think this will be enough for most of the cases, and probably could make it work in all cases with some workaround in a particular case.

> Also as you can see my implementation was a bit messy, you can probably do it better if you add such a method in the Cpp Process

Sure, not problem. I don't see it as a big difference though...

**Comment by ddiezrod** [comment]
> FYI: https://github.com/KratosMultiphysics/Kratos/blob/master/kratos/tests/test_specifications_utilities.py

I know, this is for me the second part of all this. I was just discussing this with @pooyan-dadvand, but I think this will not work in StructuralMechanics as the element type is given in mdpa, is that correct?

**Comment by loumalouomega** [comment]
> > FYI: [`master`/kratos/tests/test_specifications_utilities.py](https://github.com/KratosMultiphysics/Kratos/blob/master/kratos/tests/test_specifications_utilities.py)
> 
> I know, this is for me the second part of all this. I was just discussing this with @pooyan-dadvand, but I think this will not work in StructuralMechanics as the element type is given in mdpa, is that correct?

Depend in which point do you add the specification utilities

**Comment by loumalouomega** [comment]
> > > FYI: [`master`/kratos/tests/test_specifications_utilities.py](https://github.com/KratosMultiphysics/Kratos/blob/master/kratos/tests/test_specifications_utilities.py)
> > 
> > I know, this is for me the second part of all this. I was just discussing this with @pooyan-dadvand, but I think this will not work in StructuralMechanics as the element type is given in mdpa, is that correct?
> 
> Depend in which point do you add the specification utilities

The current behaviour iterates over the elements and conditions of the mesh, but we can add a new input interface where the elements and conditions names are provided,therefore not requiring reading any mesh. Do you want that?

**Comment by ddiezrod** [comment]
> The current behaviour iterates over the elements and conditions of the mesh, but we can add a new input interface where the elements and conditions names are provided,therefore not requiring reading any mesh. Do you want that?

Yep, that sounds good, we can do that once we are done with the processes.

**Comment by loumalouomega** [comment]
> > The current behaviour iterates over the elements and conditions of the mesh, but we can add a new input interface where the elements and conditions names are provided,therefore not requiring reading any mesh. Do you want that?
> 
> Yep, that sounds good, we can do that once we are done with the processes.

Done in #9121

**Pull request closed without merging**